_posts/2023-09-11-mcu-peripheral-forwarding.md

```diff
@@ -0,0 +1,336 @@
+---
+title: "MCU’s Peripheral Forwarding"
+description: "Mapping MCU's peripheral registers to the address space of our PC programs"
+author: evgeny
+tags: [peripherals, driver, llvm, mcu, debugger, toolchain]
+---
+
+<!-- excerpt start -->
+
+Embedded systems development often involves utilizing various microcontroller peripherals such as *CAN bus, I2C, DAC, ADC, and more*. However, when it comes to integrating these functionalities into PC applications, developers are typically required to develop firmware and communication protocols, which can add complexity to the overall software stack. This could be avoided if the chip peripheral registers were directly mapped into the address space of our PC programs.
+
+In this article, we'll explore the concept of mapping MCU’s(Chip) peripherals to your personal or embedded computer as if these peripherals were a part of the computer.
+
+There are two way to do it.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+---
+
+
+## 1st Way. Dynamic Binary Instrumentation 
+
+Imagine a scenario where you could use peripheral addresses in your PC programs just as you do in your MCU's firmware, where all address operations like store and load are dynamically hooked . This approach involves checking the operation address and, if it corresponds to peripherals, executing it on the target MCU using a debugger.
+
+![MCU Dynamic Binary Instrumentation ](/img/chip-peripheral-forwarding/intro.png)
+
+In my research, I have experimented with this implementation and documented my findings in a [**paper on medium.com**](https://medium.com/@echormonov/high-level-programming-of-mcu-periphery-from-a-host-pc-without-using-firmware-and-processor-intro-4f7dc0636b21) . 
```

Comment:
Another comment that came up -- if this article builds upon your medium one, can you call out that relationship? Otherwise, the best spot for this link would be in the resources section.